blog/2024-02-02-migrating-machines-to-xstate-v5/index.mdx

```diff
@@ -0,0 +1,563 @@
+5. [Provide concrete implementations](#5-provide-concrete-implementations)
+
+### 1. Install XState v5 and dependencies
+
+The first step will be to [install XState v5](https://stately.ai/docs/xstate#installation). You can also install library-specific dependencies if you're using XState with [React](https://stately.ai/docs/xstate-react), [Vue](https://stately.ai/docs/xstate-vue), or [Svelte](https://stately.ai/docs/xstate-svelte).
+
+If you have many state machines and would like to incrementally migrate those to v5, you can install both v4 and v5 side-by-side by [following the steps listed here](https://stately.ai/docs/migration#how-can-i-use-both-xstate-v4-and-v5). Once you've completed those steps, you'll end up with both dependencies in your `package.json`. Here's an example with libraries pegged to the latest version at the time of this writing but you can also use `npm:xstate@latest` if you prefer.
+
+```json5
+// package.json after completing the steps in the migration guide
+{
+  dependencies: {
+    xstate: '4.38.2',
+    xstate5: 'npm:xstate@5.6.0',
+    '@xstate/react': '3.2.2',
+    '@xstate/react5': 'npm:@xstate/react@4.0.3',
+  },
+}
+```
+
+:::tip
+
+Alternatively, you can also default `xstate` to v5 and use `xstate4` for v4.
+
+:::
+
+### 2. Move types to `setup()` and remove typegen
+
+One of the first migration steps, based on a breaking change, is to remove any TypeScript types in `schema` from v4. These types should then be [included under a new `types` property](https://stately.ai/docs/typescript#specifying-types) in the object passed to the new `setup()` function.
+
+<Tabs>
+<TabItem value="v5" label="XState v5">
+
+```ts
+import { setup } from 'xstate';
+
+const machine = setup({
+  types: {} as {
+    context: {
+      /* type context props */
+    },
+    events: {
+      /* type events */
+    },
+    // Type actors here
+    input:  {
+      nameOfActorSource1: {
+        output: {
+          prop1: string;
+          prop2: number;
+          /* etc */
+        };
+      };
+      nameOfActorSource2: {
+        /* type actor */
+      };
+      nameOfActorSource3: {
+        /* type actor */
+      };
+    },
+  }
+  },
+  /* implementations */
+  actions: {},
+  guards: {},
+  actors: {},
+}).createMachine({
+  /* machine config */
+});
+```
+
+</TabItem>
+
+<TabItem value="v4" label="XState v4">
+
+```ts
+// ❌ DEPRECATED
+import { createMachine } from 'xstate';
+
+const machine = createMachine({
+  ...
+  tsTypes: {} as import('./myMachine.typegen').Typegen0,
+  schema: {
+    context: {} as {
+      /* type context props */
+    },
+    events: {} as {
+      /* type events */
+    },
+    services: {} as {
+      /* type services */
+    },
+  }
+});
+```
+
+</TabItem>
+</Tabs>
+
+:::warning
+
+You can still pass `types` to `createMachine()` if you are using that on its own and if you prefer. However, if you chain `setup().createMachine()` then you must pass `types` to the call to `setup({ types: {} })`.
```

Comment:
```suggestion
You can still pass `types` to `createMachine()` if you are using that on its own and if you prefer. However, if you chain `setup().createMachine()` then you must pass `types` to the call to `setup({ types: {} })`.
```

By *that*, do you mean `types` or `createMachine()`?

Reply:
My understanding from @Andarist is that if you use `setup()` and then use the `.createMachine` returned from that function call, `setup().createMachine()` then that `.createMachine()` does not accept `types` and you are required to pass them into the `setup()`. I'll try to clarify this in the blog text. Thanks.